Add LOKI support

xakep666 opened the pull request:
add "lokiaboot" and "lokipatch" partition properties to fstab
patch partitions marked as "lokipatch" before flashing

M1cha commented:
did you test this?
Since the app is not running as root, it shouldn't be able to read the data of the aboot partition directly.

xakep666 commented:
It uses existing backup code to read aboot partition and place it dump to esp. Loki reads this file, not partition.

M1cha commented:
oh but you're writing that temporary backup to the ESP even though the partition is not being changed(it's information is just used to patch boot/recovery).

You should put the backup in the apps temporary folder instead and delete it when you're done.

xakep666 commented:
Made a dumping bootloader to getCacheDir() and removing when installation finishes

M1cha commented on app/src/main/java/org/efidroid/efidroidmanager/types/FSTabEntry.java:
```diff
@@ -105,6 +105,26 @@ public boolean isMultiboot() {
         return false;
     }
 
+    public boolean isLokiAboot() {
```
the name should be 'isLokiABoot' because A is for Android and B is for Bootloader.

M1cha commented on app/src/main/java/org/efidroid/efidroidmanager/tasks/EFIDroidInstallServiceTask.java:
```diff
@@ -161,13 +176,36 @@ private void doInstall(String updateDir) throws Exception {
+
+                String file = updateDir + "/" + entry.getName() + ".img";
+
+                //if needed, apply LOKI patch to boot and recovery partitions
```
add a space after //

M1cha commented on app/src/main/java/org/efidroid/efidroidmanager/tasks/EFIDroidInstallServiceTask.java:
```diff
@@ -161,13 +176,36 @@ private void doInstall(String updateDir) throws Exception {
+                    RootToolsEx.dd(file, entry.getBlkDevice());
+            }
+        } finally {
+            //remove bootloader dump
```
add a space after //

M1cha commented on app/src/main/java/org/efidroid/efidroidmanager/tasks/EFIDroidInstallServiceTask.java:
```diff
@@ -161,13 +176,36 @@ private void doInstall(String updateDir) throws Exception {
         }
 
-        // install
+        //dump bootloader to temporary folder
```
add a space after //

M1cha commented on app/src/main/java/org/efidroid/efidroidmanager/LokiTool.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public static boolean flashImage(PartitionLabel label, String image) {
+        //root required for flashing
```
add a space after //

M1cha commented on app/src/main/java/org/efidroid/efidroidmanager/LokiTool.java:
```diff
@@ -0,0 +1,46 @@
+
+        String mLabel;
+        PartitionLabel(String s) {
+            mLabel=s;
```
space before and after =

M1cha commented on app/src/main/java/org/efidroid/efidroidmanager/LokiTool.java:
```diff
@@ -0,0 +1,46 @@
+    private LokiTool() {}
+
+    private static native int lokiPatch(String partitionLabel, String bootloaderImage,
+                                     String inImage, String outImage);
```
this linebreak looks weird

M1cha commented on app/src/main/java/org/efidroid/efidroidmanager/LokiTool.java:
```diff
@@ -0,0 +1,46 @@
+    private static native int lokiFlash(String partitionLabel, String image);
+
+    public static boolean patchImage(PartitionLabel label, String bootloaderImage,
+                              String inImage, String outImage) {
```
this linebreak looks weird

M1cha commented on app/src/main/java/org/efidroid/efidroidmanager/tasks/EFIDroidInstallServiceTask.java:
```diff
@@ -114,6 +118,17 @@ private String downloadUpdate(String urlString) throws Exception {
     }
 
+    private static void lokiPatchAndFlash(LokiTool.PartitionLabel partition,
+                                   String bootloaderImage, String fileToPatch) throws Exception {
```
this linebreak looks weird

M1cha commented on app/src/main/java/org/efidroid/efidroidmanager/tasks/EFIDroidInstallServiceTask.java:
```diff
@@ -161,13 +176,36 @@ private void doInstall(String updateDir) throws Exception {
+                String file = updateDir + "/" + entry.getName() + ".img";
+
+                //if needed, apply LOKI patch to boot and recovery partitions
+                if (entry.isLokiPatch() &&
```
pls use nested if's to remove the duplicate condition 'isLokiPatch'

M1cha reviewed (changes requested):
you should run AndroidStudio's auto-formatting tool on all the files which would make the changes I've requested much easier.

xakep666 commented:
Improved encapsulation in "partitionlabel" enum, fixed formatting

M1cha commented on app/src/main/java/org/efidroid/efidroidmanager/LokiTool.java:
```diff
@@ -0,0 +1,49 @@
+        // root required for flashing
+        boolean isRootAccess = RootTools.isAccessGiven();
+        if (!isRootAccess) {
+            Log.e(TAG, "flashImage: cannot get root privileges");
```
Sry that I missed this previously, but why are you using the volley logtag here?
Also why are you checking for root in first place?
1) it has already been done at this point
2) the call to lokiFlash doesn't need root.

xakep666 replied:
About tag: autocompleted at "loge"

xakep666 replied:
Root required because of native library writes new boot and recovery parts

M1cha replied:
and where does this happen? NDK-code does not run as root and cannot write to any partitions.

xakep666 replied:
https://github.com/efidroid/modules_loki/blob/master/loki_flash.c#L133

M1cha replied:
well unless your ROM uses super insecure permissions and selinux is disabled this line will always fail.

xakep666 replied:
You mean I have to use RootToolsEx.dd() to flash patched parts

M1cha replied:
either that or you have to compile loki as a statically linked binary like busybox and run it as root.
So, didn't you test your change yet? Because such a mistake should have been noticed.

xakep666 replied:
I have no device to test this tool. Also, at first idea was compiling loki_tool. Now I think, maybe just place precompiled binary into "assets" and call it

M1cha reviewed (changes requested):
see my comments

xakep666 commented:
Rewrote using separate binary (runs on my device)

M1cha commented on app/src/main/java/org/efidroid/efidroidmanager/LokiTool.java:
```diff
@@ -0,0 +1,73 @@
+        try {
+            InputStream is = null;
+            try {
+                is = context.getAssets().open("armeabi-v7a/loki_tool_static");
```
this binary isn't statically linked. Linking it dynamically actually isn't such a bad idea because it makes the file size much smaller. But what about compatibility? we need to support 4.0+.
Also, is it possible to build binaries with gradle+ndk? The only reason I'm using busybox prebuilt is that it doesn't work that well when compiled using the ndk and still needs to be built using autoconf anyway.

Either way, we don't include the link-type in the filename.

xakep666 replied:
I didn't find way to build and add binary to apk using gradle.
I built it with ndk-build which uses Android.mk included in sources.

M1cha commented on app/src/main/java/org/efidroid/efidroidmanager/LokiTool.java:
```diff
@@ -0,0 +1,73 @@
+            abis.add(Build.CPU_ABI2);
+        }
+        // do not initialize on non-arm cpu
+        if (!abis.contains("armeabi-v7a")) return;
```
You should rather make the original code for busybox reusable, also don't limit this to ARM.
If you do this you can also keep this code inside RootToolsEx since this implementation shouldn't have to care about where the binary comes from.

xakep666 replied:
This tool is usable only for arm-based phones. 
See https://github.com/efidroid/modules_loki/blob/master/loki_patch.c#L30

M1cha reviewed (changes requested):
see my comments

xakep666 committed:
Add possibility to patch image before flashing

FSTabEntry: getFfMgrFlags() returns list of flags. Rewrite isMultiboot(), isUEFI(), getESP() using it

Move ABIs retrieving code to separate method getABIs() in Util

Add universal image patching interface
Add image patching (if needed) before flashing.
Add LOKI as git submodule
Add LOKI image patcher using NDK
